src/gtkspellcheck/spellcheck.py

```diff
@@ -149,32 +151,121 @@ def exists(self, language):
-            return self.iter.inside_word()
+            return self._iter_worker.inside_word(self.iter)
 
         @property
         def word(self):
             start = self.iter
-            if not start.starts_word():
-                start.backward_word_start()
+            if not self._iter_worker.starts_word(start):
+                self._iter_worker.backward_word_start(start)
             end = self.iter
-            if end.inside_word():
-                end.forward_word_end()
+            if self._iter_worker.inside_word(end):
+                self._iter_worker.forward_word_start(end)
             return start, end
 
         def move(self, location):
             self._buffer.move_mark(self._mark, location)
 
+    class _IterWorker:
+        def __init__(self, extra_word_chars):
+            self._extra_word_chars = extra_word_chars
+
+        def is_extra_word_char(self, loc):
+            # Language extra chararacters should also be processed once Enchant's
+            # enchant_dict_get_extra_word_characters is exposed in PyEnchant
+
+            char = loc.get_char()
+            return char != "" and char in self._extra_word_chars
+
+        def inside_word(self, loc):
+            if loc.inside_word():
+                return True
+            elif self.starts_word(loc):
+                return True
+            elif loc.ends_word() and not self.ends_word(loc):
+                return True
+            else:
+                return False
+
+        def starts_word(self, loc):
+            if loc.starts_word():
+                if loc.is_start():
+                    return True
+                else:
+                    tmp = loc.copy()
+                    tmp.backward_char()
+                    return not self.is_extra_word_char(tmp)
+            else:
+                return False
+
+        def ends_word(self, loc):
+            if loc.ends_word():
+                if loc.is_end():
+                    return True
+                else:
+                    tmp = loc.copy()
+                    tmp.forward_char()
+                    return not self.is_extra_word_char(tmp)
+            else:
+                return False
+
+        def forward_word_end(self, loc):
+            def move_through_extra_chars():
+                moved = False
+                while self.is_extra_word_char(loc):
+                    moved = True
+                    if not loc.forward_char():
+                        break
+                return moved
+
+            tmp = loc.copy()
+            tmp.backward_char()
+            loc.forward_word_end()
+            while move_through_extra_chars():
+                if loc.is_end() or not loc.inside_word() or not loc.forward_word_end():
+                    break
+
+            tmp = loc.copy()
+            tmp.backward_char()
+
+        def backward_word_start(self, loc):
+            def move_through_extra_chars():
+                tmp = loc.copy()
+                tmp.backward_char()
```

Comment:
Why do we need this?

In `forward_word_end` this is not part of the `move_through_extra_chars` function.

Reply:
Because we're iterating back over the characters and we move the temporary iter to the character we want to check? Going forwards this isn't necessary as the iter is already pointing at the character. The difference in how we end up specifying the envelope?

Reply:
Ah, I see, makes perfect sense. 👌